function_test.go

```diff
@@ -135,3 +136,34 @@ func Test_Interpolate(t *testing.T) {
 		t.Fatalf("expected envs with a nil value to not be included in interpolation result")
 	}
 }
+
+// TestFunction_MarshallingError check that the correct error gets reported back to the
+// user if the function that is being loaded is failing marshalling and cannot be migrated
+func TestFunction_MarshallingError(t *testing.T) {
+	root := "testdata/testFunctionMarshallingError"
+
+	// Load the function to see it fail with a marshalling error
+	_, err := fn.NewFunction(root)
+	if err != nil {
+		if !strings.Contains(err.Error(), "line 2: cannot unmarshal !!seq into string") {
+			t.Fatalf("expected unmarshalling error")
+		}
+
+	}
```

Comment:
A minor nit here; and there are indeed a few ways to go about this (ranging from simple to complete);  but any time we want to test that a _specific_ error is being returned, it's better to use a typed error; creating an error type for the error to be tested.  This is not only more clear and less brittle than a string test, but also will greatly ease the potential future pain of internationalization/localization of text later:

For example, assuming we have an error type:
```go
type ErrUnmarshalling struct{
  Err error
}
func (e ErrUnmarhsalling) Error() string {
   return fmt.Sprintf("unable to read function: %s", e.Err)
}
```
And a reporting of that error somewhere:
```go
if err = yaml.Unmarshal(bb, &f0); err != nil {
		return f, ErrUnmarshalling{err}
}
```
Then the tests can be:
```go
if err := fn.NewFunction(root); !errors.Is(err, ErrUnmarshalling) {
  t.Fatalf("expected unmarshalling error not received")
}
```

Reply:
I am all in for this.. I will work on that fix after this large PR is merged